Add support for two tokens in erc to native

patitonar opened the pull request:
Closes #318

akolotov commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -71,7 +79,13 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
-        return erc20token().transfer(_recipient, amount);
+        bool res = erc20token().transfer(_recipient, amount);
+
+        if (AddressUtils.isContract(halfDuplexErc20token()) && tokenBalance(halfDuplexErc20token()) > 0) {
```
do you think it make sense to check if it is a contract for the case when we hardcode the return value of the method `halfDuplexErc20token`? When we hardcode the value we know that it is the token contract.

patitonar replied:
You are correct, the check is not needed. Fixed https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/055bcb62a794ed913dfec10b6558e3b57a314151

akolotov commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -102,4 +116,100 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
+        return _token.balanceOf(address(this));
+    }
+
+    function migrationContract() internal pure returns (IScdMcdMigration) {
```
we don't need to specify the address of the migration contract in `migrateToMCD` anymore, do we?

patitonar replied:
Thanks, fixed https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/360dcaaafefa2069f363becae01c3a71bf4a3a8c

akolotov commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -102,4 +116,100 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
+        return IScdMcdMigration(0xc73e0383F3Aff3215E6f04B0331D58CeCf0Ab849);
+    }
+
+    function swapTokens() public {
```
Can this method be used in `migrateToMCD`?

patitonar replied:
Updated `migrateToMCD` to call `swapTokens`

akolotov commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -102,4 +116,100 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
+        return ERC20(0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359);
+    }
+
+    function setMinHDTokenBalance(uint256 _minBalance) external onlyOwner {
```
Let's set some initial value (e.g. `10 ether`) as part of the `migrateToMCD` call

patitonar replied:
Added initial value https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/360dcaaafefa2069f363becae01c3a71bf4a3a8c

patitonar commented:
Tested the implementation with [MakerDAO contracts in Kovan testnet](https://changelog.makerdao.com/releases/kovan/0.2.17/index.html) which are identical to the deployed in mainnet.

Sai Address: 0xc4375b7de8af5a38a93548eb8453a498222c4ff2
Dai Address: 0x4f96fe3b7a6cf9725f59d353f723c1bdb64ca6aa
ScdMcdMigration Address: 0x411b2faa662c8e3e5cf8f01dfdae0aee482ca7b0
Sai Top: 0x5f00393547561da3030ebf30e52f5dc0d5d3362c (from https://developer.makerdao.com/dai/1/api/)

Foreign Bridge Address: 0xb644F2c13D1e2410d224825991455fCBE95ACcd9

List of tx examples
- migrateToMCD [0x1470e2434ea73e3b4fcb10a8007b2f3def9e107746c9f615d9aea86dd3774c7f](https://blockscout.com/eth/kovan/tx/0x1470e2434ea73e3b4fcb10a8007b2f3def9e107746c9f615d9aea86dd3774c7f)
- swapTokens [0x3e2b6b891aec411f378cf60fc04ebe6554f19ec14827ccc9d559e1ae8b183e02](https://blockscout.com/eth/kovan/tx/0x3e2b6b891aec411f378cf60fc04ebe6554f19ec14827ccc9d559e1ae8b183e02)
- relayTokens sai [0x34ecac9a85cc1d44b86e22d70c2906294ab08d4f124073eb9296c4c904cd8eee](https://blockscout.com/eth/kovan/tx/0x34ecac9a85cc1d44b86e22d70c2906294ab08d4f124073eb9296c4c904cd8eee)
- relayTokens dai [0x1b1c1141ec38ba9e9321e05c91f34a9d7e610812954bf75ff5f868095484a4ee](https://blockscout.com/eth/kovan/tx/0x1b1c1141ec38ba9e9321e05c91f34a9d7e610812954bf75ff5f868095484a4ee)

akolotov commented:
> Tested the implementation with [MakerDAO contracts in Kovan testnet](https://changelog.makerdao.com/releases/kovan/0.2.17/index.html) which are identical to the deployed in mainnet.

thanks!

akolotov commented on contracts/upgradeable_contracts/BaseBridgeValidators.sol:
```diff
@@ -102,4 +102,23 @@ contract BaseBridgeValidators is InitializableBridge, Ownable {
         addressStorage[keccak256(abi.encodePacked("validatorsList", _prevValidator))] = _validator;
     }
+
+    function isValidatorDuty(address _validator) external view returns (bool) {
```
should we increase a version in `getBridgeValidatorsInterfacesVersion`?

patitonar replied:
I think we should. Updated version to `2.3.0` https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/be3cd335b2bf2f4077759dfc0996e52e00b367a7

akolotov commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -83,23 +97,113 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
+        return IScdMcdMigration(0xc73e0383F3Aff3215E6f04B0331D58CeCf0Ab849);
+    }
+
+    function swapTokens() public {
```
should we increase the version returned by `getBridgeInterfacesVersion`?

patitonar replied:
Yes, it makes sense. Updated to `2.5.0` https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/c2c2c4cc2b5308b1fbe5a7bdd872d5238e297636

patitonar commented on contracts/upgradeable_contracts/erc20_to_native/ForeignBridgeErcToNative.sol:
```diff
@@ -83,23 +97,113 @@ contract ForeignBridgeErcToNative is BasicForeignBridge, ERC20Bridge, OtherSideB
     }
 
-    function migrateToMCD(address _migrationContract) external onlyOwner {
+    function migrateToMCD() external onlyOwner {
```
Updated the method https://github.com/poanetwork/tokenbridge-contracts/pull/319/commits/251f9495578ca63202bc731068dc8064c04fdded

patitonar commented:
Tested the upgrade with full changes following the steps mentioned [here](https://hackmd.io/@1oqceZGlR6WSWAv2TTWJeA/Bk76GRPpS). Everything worked as expected and no issues were found.

Here are the new contract implementations:
* **[xDai] HomeBridgeErcToNative**: no migration method required. Gist: https://gist.github.com/patitonar/a19287408dfa878d7be1dca95835625f
* **[mainnet] BridgeValidators**: added migration method to create the list of validators. Gist: https://gist.github.com/patitonar/fe4888e786e9b6cdbfaeedaf8fe99b78
```javascript
function upgradeToV230() public {
    bytes32 upgradeStorage = 0x2e8a0420c66b3c19df2ac1cd4ffadb5637da547c6fd49388c7fe28dc8c13a8dd; // keccak256(abi.encodePacked('isUpgradedToV230'))
    require(!boolStorage[upgradeStorage]);
    address validator1 = 0x9ad83402C19Af24F76afa1930a2B2EEC2F47A8C5;
    address validator2 = 0x4D1c96B9A49C4469A0b720a22b74b034EDdFe051;
    address validator3 = 0x491FC792e78CDadd7D31446Bb7bDDef876a69AD6;
    address validator4 = 0xc073C8E5ED9Aa11CF6776C69b3e13b259Ba9F506;

    setNextValidator(F_ADDR, validator1);
    setNextValidator(validator1, validator2);
    setNextValidator(validator2, validator3);
    setNextValidator(validator3, validator4);
    setNextValidator(validator4, F_ADDR);
    boolStorage[upgradeStorage] = true;
}
```
* **[mainnet] ForeignBridgeErcToNative**: added migration method for missing parameters. Gist: https://gist.github.com/patitonar/c0b77a59eceb76aad05e010c40569d47
```javascript
function upgradeToV250() public {
    bytes32 upgradeStorage = 0x221181d03d3c423368b412861518cd98276a115f4c2d9594ab6697d90fc99c19; // keccak256(abi.encodePacked('isUpgradedToV250'))
    require(!boolStorage[upgradeStorage]);
    uintStorage[DAILY_LIMIT] = 100000 ether;
    emit DailyLimitChanged(100000 ether);
    uintStorage[MIN_PER_TX] = 5000000000000000;
    _setBridgeContractOnOtherSide(0x7301CFA0e1756B71869E93d4e4Dca5c7d0eb0AA6);
    boolStorage[upgradeStorage] = true;
}
```

I tested all steps using the strategy option 1 on Sokol and Kovan, using the set of contracts provided by makerDAO on Kovan. Except for step `9. Test Emergency Shutdown` which was not possible to generate such a scenario.

I also tested upgrading in foreign side `BridgeValidators` and `ForeignBridgeErcToNative` with option 2 (using `ganache-cli` to fork the networks). I was able to deploy the new implementations and upgrade the contracts. I checked the state was not lost and new variables were correctly set. But I wasn't able to test the ui, oracle, and monitor in this case, because the forked local network generated by ganache was slow and sometimes it just stopped working when received a lot of requests, especially when trying to get events. Also, I wasn't able to deploy the new implementation for `HomeBridgeErcToNative` in the forked xDai, probably because of some issues with `ganache-cli`. I tested deploying the same contract directly in xDai and it worked correctly there.
